### PR TITLE
Return registration errors to client

### LINKED ISSUE #264: Machine registration failures should propagate to the journal on the machine being registered
I tried to deploy two machines following the quickstart guide, with registration.yaml specifying [machineName: my-machine](https://github.com/rancher/elemental-docs/blob/661f0ba3d1cc8aff8d43b8908bb9a43a3458a7c3/examples/quickstart/registration.yaml#L17). The first machine deployed fine, the second one failed. While I was deploying the second machine, I ran `journalctl -f -t elemental` as suggested on the console, and it endlessly repeated the following:

```
Nov 24 05:26:13 rancher-2909 elemental[1380]: time="2022-11-24T05:26:13Z" level=debug msg="start TPM attestation"
Nov 24 05:26:13 rancher-2909 elemental[1380]: time="2022-11-24T05:26:13Z" level=info msg="TPM attestation successful"
Nov 24 05:26:13 rancher-2909 elemental[1380]: time="2022-11-24T05:26:13Z" level=debug msg="send SMBIOS data"
Nov 24 05:26:13 rancher-2909 elemental[1380]: time="2022-11-24T05:26:13Z" level=debug msg="get elemental configuration"
Nov 24 05:26:13 rancher-2909 elemental[1380]: time="2022-11-24T05:26:13Z" level=error msg="failed to register machine inventory: read elemental configuration: websocket: close 1006 (abnormal closure): unexpected EOF"
```

It wasn't until I figured out to run `kubectl logs elemental-operator-775d455cc8-gnmbq -n cattle-elemental-system` that I understood the problem:

```
time="2022-11-24T05:30:46Z" level=error msg="MachineInventory creation failed: machineinventories.elemental.cattle.io \"my-machine\" already exists"
```

Is it possible to somehow propagate errors like this back to the machine being registered, so that they're visible in the `journalctl` output, rather than just seeing `unexpected EOF`?

**Environment:**

- Elemental release version (use `cat /etc/os-release`): 5.3:0.2.0-24.27
- Rancher version: 2.6.9
- Kubernetes version (use `kubectl version`): v1.24.8+k3s1
- Cloud provider or hardware configuration: libvirt VMs on my Tumbleweed desktop system

Also the elemental operator version is 0.6.3, dug out of `kubectl logs -p -n cattle-elemental-system -l app=elemental-operator`. Is there a better way to get that?

### PR DESCRIPTION
**Pull request by frelon**
Introduces two new message-types (MsgError and MsgConfig).

MsgError is sent when an error is encountered during the registration process.

MsgConfig is used to send the elemental configuration to the client, before this was just a raw message with no type so we need to check in the server if the client supports the message, otherwise fallback to the raw message.

Example journal from trying to register a second machine with name "my-machine":

```
...
Dec 16 08:42:51 rancher-18652 elemental[1443]: time="2022-12-16T08:42:51Z" level=info msg="Negotiated protocol version: 8"
Dec 16 08:42:51 rancher-18652 elemental[1443]: time="2022-12-16T08:42:51Z" level=info msg="Send SMBIOS data"
Dec 16 08:42:51 rancher-18652 elemental[1443]: time="2022-12-16T08:42:51Z" level=info msg="Send system data"
Dec 16 08:42:51 rancher-18652 elemental[1443]: time="2022-12-16T08:42:51Z" level=info msg="Get elemental configuration"
Dec 16 08:42:51 rancher-18652 elemental[1443]: time="2022-12-16T08:42:51Z" level=debug msg="Got configuration response: Error"
Dec 16 08:42:51 rancher-18652 elemental[1443]: time="2022-12-16T08:42:51Z" level=error msg="failed to register machine inventory: MachineInventory creation failed: failed to create machine inventory fleet-default/my-machine: machineinventories.elemental.cattle.io \"my-machine\" already exists"

```

Fixes #264 

Signed-off-by: Fredrik Lönnegren <fredrik.lonnegren@suse.com>

### CODE REVIEW

**Review by fgiudici — APPROVED** [review]
Cool, we can get meaningful error messages on the client now!

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -84,3 +84,7 @@ type Config struct {
 	// +optional
 	CloudConfig map[string]runtime.RawExtension `json:"cloud-config,omitempty" yaml:"cloud-config,omitempty"`
 }
+
+type ErrorMessage struct {
```
Comment (fgiudici) [inline api/v1beta1/types.go:88]: 🤔 not sure we should putting this here, in the operator controllers api. This is used only in the registration process.
What about moving it in the pkg/register/websocket.go file? After all it is related only to the websocket communication?

**Review by alexander-demicev — APPROVED** [review]

File: api/v1beta1/types.go (hunk 1 above)
Comment (frelon) [reply api/v1beta1/types.go:88]: Yep, that's a better idea! I will change that and also try to add some tests :+1:

File: pkg/register/register.go [hunk 2]
```diff
@@ -100,11 +102,29 @@ func Register(url string, caCert []byte, smbios bool, emulateTPM bool, emulatedS
...
+			return nil, errors.Wrap(err, "unable to unmarshal error-message")
+		}
+
+		return nil, errors.New(msg.Message)
+
+	case MsgConfig:
+		return data, nil
+
+	default:
+		return nil, fmt.Errorf("Unknown response message-type: %s", msgType)
```
Comment (fgiudici) [inline pkg/register/register.go:126]: nit: here you will get either an unexpected known message (that will expand to the corresponding string) or a code not know that will expand to "Unknown".
```suggestion
		return nil, fmt.Errorf("unexpected response message: %s", msgType)
```

File: pkg/server/register.go [hunk 3]
```diff
@@ -395,6 +403,51 @@ func (i *InventoryServer) serveLoop(conn *websocket.Conn, inventory *elementalv1
...
+	if protoVersion == register.MsgUndefined {
+		logrus.Warn("Detected old elemental-register client: cloud-config data may not be applied correctly")
+	}
+
+	return nil
+}
+
+// writeError reports back an error to the client if the negotiated protocol
+// version supports it.
+func (i *InventoryServer) writeError(conn *websocket.Conn, protoVersion register.MessageType, errorMessage register.ErrorMessage) error {
```
Comment (fgiudici) [inline pkg/server/register.go:437]: Nice, the writeError function!
nit: we don't need the InventoryServer receiver argument here
```suggestion
func writeError(conn *websocket.Conn, protoVersion register.MessageType, errorMessage register.ErrorMessage) error {
```

**Review by fgiudici — APPROVED** [review]
Pretty nice improvement, love the new tests! 🥳 
Just a couple of nits that we can address after merging this PR.

File: pkg/server/register.go (hunk 3 above)
Comment (frelon) [reply pkg/server/register.go]: Nice, removing this in a new commit :+1:

**Comment by davidcassany** [comment]
@frelon cool, thanks for the additional tests! :+1:

**Pull request merged**